### PR TITLE
Revamp assessment.js library

### PR DESCRIPTION
**Pull request by jonatanschroeder**
* Convert to ESM notation;
* Use new query functions;
* Promisify functions still in callback mode.

### CODE REVIEW

File: apps/prairielearn/src/lib/assessment.js [hunk 1]
```diff
@@ -40,482 +42,417 @@ const InstanceLogSchema = z.object({
...
-   */
-  async selectAssessmentInstanceLog(assessment_instance_id, include_files) {
-    return sqldb.queryValidatedRows(
-      sql.assessment_instance_log,
-      { assessment_instance_id, include_files },
-      InstanceLogSchema,
+
+    if (_(content.grading).has('feedback') && !_(content.grading.feedback).isObject()) {
+      throw error.makeWithData('invalid grading.feedback', { content: content });
+    }
```
Comment (jonatanschroeder) [inline apps/prairielearn/src/lib/assessment.js:316]: I'm considering having some of this validation be moved to a Zod schema instead, but I believe that should be done in a separate PR that focuses on that alone and does proper testing, and that is done once this file (and externalGrader.js, and possibly externalGraderCommon.js) are fully converted to TS.

File: apps/prairielearn/src/pages/studentAssessmentInstanceHomework/studentAssessmentInstanceHomework.js [hunk 2]
```diff
@@ -17,23 +17,27 @@ const sql = sqldb.loadSqlEquiv(__filename);
 
 const ensureUpToDate = (locals, callback) => {
   debug('ensureUpToDate()');
-  assessment.update(locals.assessment_instance.id, locals.authn_user.user_id, (err, updated) => {
-    if (ERR(err, callback)) return;
+  util.callbackify(assessment.update)(
+    locals.assessment_instance.id,
+    locals.authn_user.user_id,
+    (err, updated) => {
```
Comment (jonatanschroeder) [inline apps/prairielearn/src/pages/studentAssessmentInstanceHomework/studentAssessmentInstanceHomework.js:23]: This is the only reference to this function. I didn't want to convert this to async in this PR as the code is a bit complex to do just a partial change, but I kept the reference using `util.callbackify` so that a future revamp of this code has an easier time converting this function.

File: apps/prairielearn/src/lib/assessment.js [hunk 3]
```diff
@@ -36,486 +38,414 @@ const InstanceLogSchema = z.object({
...
- * Assessment module.
- * @module assessment
+ * Check that an assessment_instance_id really belongs to the given assessment_id
+ *
+ * @param {string} assessment_instance_id - The assessment instance to check.
+ * @param {string} assessment_id - The assessment it should belong to.
  */
+export async function checkBelongsAsync(assessment_instance_id, assessment_id) {
+  if (
+    (await sqldb.queryRow(
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/assessment.js:48]: `queryRow` will throw its own error if a row is not found. I believe you want `queryOptionalRow` here instead.

File: apps/prairielearn/src/lib/assessment.js [hunk 4]
```diff
@@ -36,486 +38,414 @@ const InstanceLogSchema = z.object({
...
-  gradeAssessmentInstance(
+  ]);
+  return result.rows[0].assessment_instance_id;
+}
+
+/**
+ * Add new questions to the assessment instance and regrade it if necessary.
+ *
+ * @param {string} assessment_instance_id - The assessment instance to grade.
+ * @param {string} authn_user_id - The current authenticated user.
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/assessment.js:114]: Consider adding an `@returns` JSDoc here to document that this returns a boolean that indicates if the assessment instance was updated or not.

File: apps/prairielearn/src/pages/studentAssessmentExam/studentAssessmentExam.js [hunk 5]
```diff
@@ -1,4 +1,3 @@
-const ERR = require('async-stacktrace');
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/studentAssessmentExam/studentAssessmentExam.js:1]: 🙌 it's so good to see these imports disappearing!

**Review by nwalters512 — COMMENTED** [review]
Will approve after discussion of `queryRow -> queryOptionalRow`.

File: apps/prairielearn/src/lib/assessment.js (hunk 4 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/lib/assessment.js:117]: Done in fa84191bfbb4c81d5657adec0cc55479cfbc5c05. I also added this in all functions.

File: apps/prairielearn/src/lib/assessment.js (hunk 3 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/lib/assessment.js]: Fixed in 1daba6c17c71e2bde978d382c2551ddcbe09a027.

File: apps/prairielearn/src/lib/assessment.js [hunk 6]
```diff
@@ -36,486 +38,429 @@ const InstanceLogSchema = z.object({
...
+    { assessment_instance_id, include_files },
+    InstanceLogSchema,
+  );
+}
+
+/**
+ *
+ * @param {string} assessment_instance_id
+ * @param {boolean} include_files
+ * returns {Promise<sqldb.CursorIterator<z.infer<typeof InstanceLogSchema>>>} (Not using jsdoc since CursorIterator is not exported yet)
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/assessment.js:458]: One of us should follow up to export this type.

**Review by nwalters512 — APPROVED** [review]

File: apps/prairielearn/src/lib/assessment.js (hunk 6 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/lib/assessment.js:458]: Will do when I convert it to TS.
